smart_sa/intervention/models.py

```diff
@@ -869,6 +871,226 @@ def load_game_vars(self, data):
             for key in gv.keys():
                 self.save_game_var(key, gv[key])
 
+    def activity_sort(self, string):
+        match = re.search(r"\d{1,3}", string)
+        return int(match.group()) if match else 0
+
+    def get_completed_activities(self):
+        activities = []
+        for i in range(1, ClientSession.objects.count()):
+            activities.append(["Activity %d: %s" % (
+                pa.activity.index(), pa.activity.long_title)
+                for pa in self.participantactivity_set.all()
+                if pa.status == u'complete' and
+                pa.activity.clientsession.index() == i])
+
+        for i in range(len(activities)):
+            activities[i].sort(key=self.activity_sort)
+
+        return activities
+
+    def session_duration(self, session):
+        """ approximate number of minutes participant spent
+        on a given session
+
+        since we just care about a total, not whether it was all
+        continuous or not, we can "cheat" and take advantage
+        of the fact that the front end logs every minute and just
+        count the number of relevant visit logs.
+
+        That can go wrong if they are flying through and visiting
+        more than one page a minute; we'd get a higher number from
+        the count than the actual total time. So we compare our
+        count to the absolute difference between the earliest and
+        latest timestamps we have and take the min.
+
+        This can still be inaccurate if they flew through parts
+        and then had gaps (computer shut off, logged out, etc)
+        in other parts. So a proper run through merging timestamps
+        that are less than a minute apart into runs would be ideal.
+        But this should get us pretty close without too much craziness.
+        Since the logging is pretty rough at best, this is probably
+        good enough, at least for now.
+        """
+        timestamps = self.relevant_timestamps(session)
+        simple_count = len(timestamps)
+
+        if simple_count < 2:
+            # zero or 1 timestamps logged, so this is as
+            # accurate as we can get
+            return simple_count
+
+        # otherwise, double check against absolute time interval
+        timestamps.sort()
+        earliest = timestamps[0]
+        latest = timestamps[-1]
+        delta = latest - earliest
+        if (delta.seconds / 60.0) < simple_count:
+            return delta.seconds / 60
+        else:
+            return simple_count
+
+    def relevant_timestamps(self, session):
+        """ all the timestamps for a session and activities in that session """
+        return (self.session_timestamps(session) +
+                self.session_activity_timestamps(session))
+
+    def session_timestamps(self, session):
+        sv_set = self.sessionvisit_set.all()
+        if not sv_set:
+            return []
+        return [sv.logged for sv in sv_set
```

Comment:
Could this function (and a few similar functions) be simplified by just filtering on the passed session.id and returning the queryset?